podutils/env.go

```diff
@@ -0,0 +1,598 @@
+	ResolveEnvVarExpansions(pod)
+	UniqifyEnvVars(pod)
+	return nil
+}
+
+// Resolves all EnvFrom and Env ConfigMap environment variable
+// reference in the pod. Any optional references that don't have a
+// valid configmap will be left unresolved in the environment.
+func ResolveConfigMapRefs(ctx context.Context, pod *corev1.Pod, configMapLister corev1listers.ConfigMapLister, recorder record.EventRecorder) error {
+	configMaps := make(map[string]*corev1.ConfigMap)
+
+	return forEachContainerWithError(pod, func(container *corev1.Container) error {
+		if err := resolveContainerConfigMapRefs(ctx, pod, container, configMapLister, recorder, configMaps); err != nil {
+			return err
+		}
+		return nil
+	})
+}
+
+// Resolves EnvFrom and Env ConfigMap environment variable references
+// in a container.
+func resolveContainerConfigMapRefs(ctx context.Context, pod *corev1.Pod, container *corev1.Container, configMapLister corev1listers.ConfigMapLister, recorder record.EventRecorder, configMaps map[string]*corev1.ConfigMap) error {
+	envFromVals := make([]corev1.EnvVar, 0)
+	for i := range container.EnvFrom {
+		if container.EnvFrom[i].ConfigMapRef != nil {
+			envFrom := container.EnvFrom[i]
+			cm := envFrom.ConfigMapRef
+			name := cm.Name
+			configMap, ok := configMaps[name]
+			if !ok {
+				var err error
+				optional := cm.Optional != nil && *cm.Optional
+				configMap, err = configMapLister.ConfigMaps(pod.Namespace).Get(name)
+				if err != nil {
+					if errors.IsNotFound(err) && optional {
+						recorder.Eventf(pod, corev1.EventTypeWarning, ReasonOptionalConfigMapNotFound, "configmap %q not found", name)
+						continue
+					}
+					if errors.IsNotFound(err) {
+						recorder.Eventf(pod, corev1.EventTypeWarning, ReasonMandatoryConfigMapNotFound, "configmap %q not found", name)
+						return fmt.Errorf("configmap %q not found", name)
+					}
+					recorder.Eventf(pod, corev1.EventTypeWarning, ReasonFailedToReadMandatoryConfigMap, "failed to read configmap %q", name)
+					return fmt.Errorf("failed to fetch configmap %q: %v", name, err)
+				}
+				configMaps[name] = configMap
+			}
+
+			invalidKeys := []string{}
+			for k, v := range configMap.Data {
+				if len(envFrom.Prefix) > 0 {
+					k = envFrom.Prefix + k
+				}
+				if errMsgs := apivalidation.IsEnvVarName(k); len(errMsgs) != 0 {
+					invalidKeys = append(invalidKeys, k)
+					continue
+				}
+				envFromVals = append(envFromVals, corev1.EnvVar{
+					Name:  k,
+					Value: v,
+				})
+			}
+			if len(invalidKeys) > 0 {
+				sort.Strings(invalidKeys)
+				recorder.Eventf(pod, corev1.EventTypeWarning, ReasonInvalidEnvironmentVariableNames, "Keys [%s] from the EnvFrom configMap %s/%s were skipped since they are considered invalid environment variable names.", strings.Join(invalidKeys, ", "), pod.Namespace, name)
+			}
+		}
+	}
+
+	for i := range container.Env {
+		if container.Env[i].ValueFrom != nil && container.Env[i].ValueFrom.ConfigMapKeyRef != nil {
+			envVar := container.Env[i]
```

Comment:
I recommend using a code-style like this:
```
if container.Env[i].ValueFrom != nil {
    continue
}

if container.Env[i].ValueFrom.ConfigMapKeyRef != nil {
    continue
}
```

Reply:
I like it since this gets rid of a level of indentation. However I prefer the coupling of the nested checks into a single statement using a short circuit `||`. That shows that the checks for `nil` are related, reducing a bit of the cognitive overhead as well as reducing line count:

```go
if container.Env[i].ValueFrom == nil ||
    container.Env[i].ValueFrom.ConfigMapKeyRef == nil {
    continue
}
```